src/fairseq2/recipes/lm/instruction_finetune.py

```diff
@@ -296,12 +295,21 @@ def load_instruction_finetuner(
     """Load a :class:`Trainer` for language model instruction-finetuning."""
     wall_watch = Stopwatch(start=True)
 
-    root_gang = resolve(Gang)
+    root_gang, gangs = setup_gangs(
+        log, tp_size=config.tensor_parallel_size, monitored=config.monitored_gang
+    )
 
-    dp_gang = resolve(Gang, key="dp")  # data
-    tp_gang = resolve(Gang, key="tp")  # tensor
+    dp_gang = gangs["dp"]  # data
+    tp_gang = gangs["tp"]  # tensor
 
-    checkpoint_manager = resolve(CheckpointManager)
+    checkpoint_manager = FileCheckpointManager(
+        output_dir.joinpath("checkpoints"), root_gang, dp_gang=dp_gang, tp_gang=tp_gang
+    )
+
+    if config.resume_checkpoint_dir is not None:
+        default_asset_store.metadata_providers.append(
+            CheckpointModelMetadataProvider(config.resume_checkpoint_dir)
+        )
```

Comment:
This code snippet shows up in multiple recipes. I wonder if it make sense to abstract `default_asset_store` away from recipe creators and put it as a function to add the ckpt provider into `recipes/utils/asset.py`?

Reply:
Same as above. This is in a common code base now to reduce the boilerplate code in recipes.